cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConcurrentPartitionConsumer.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.dataset2.lib.partitioned;
+
+import co.cask.cdap.api.dataset.lib.Partition;
+import co.cask.cdap.api.dataset.lib.PartitionDetail;
+import co.cask.cdap.api.dataset.lib.PartitionKey;
+import co.cask.cdap.api.dataset.lib.PartitionedFileSet;
+import co.cask.cdap.api.dataset.lib.partitioned.ConsumerConfiguration;
+import co.cask.cdap.api.dataset.lib.partitioned.PartitionConsumer;
+import co.cask.cdap.api.dataset.lib.partitioned.StatePersistor;
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link PartitionConsumer} that supports multiple instances consuming the same set of partitions by using a
+ * working set of partitions, and keeping track of their progress state during processing of those partitions.
+ */
+public class ConcurrentPartitionConsumer implements PartitionConsumer {
+  private final PartitionedFileSet partitionedFileSet;
+  private final ConsumerConfiguration configuration;
+  private final StatePersistor statePersistor;
+
+  /**
+   * Creates an instance of ConcurrentPartitionConsumer which consumes partitions,
+   *
+   * @param partitionedFileSet the PartitionedFileSet to consume from
+   * @param statePersistor defines how the state of the PartitionConsumer will be maintained
+   * @param configuration the PartitionedConsumerConfiguration, defining parameters of consumption
+   */
+  public ConcurrentPartitionConsumer(PartitionedFileSet partitionedFileSet,
+                                     StatePersistor statePersistor, ConsumerConfiguration configuration) {
+    this.partitionedFileSet = partitionedFileSet;
+    this.configuration = configuration;
+    this.statePersistor = statePersistor;
+  }
+
+  @Override
+  public List<PartitionDetail> consumePartitions() {
+    return consumePartitions(Integer.MAX_VALUE);
+  }
+
+  @Override
+  public List<PartitionDetail> consumePartitions(int limit) {
+    ConsumerWorkingSet workingSet = readState();
+
+    ConsumerWorkingSetOperations consumerWorkingSetOperations =
+      new ConsumerWorkingSetOperations(configuration, partitionedFileSet, workingSet);
+    List<PartitionDetail> partitionsToConsume = consumerWorkingSetOperations.consumePartitions(limit);
+
+    statePersistor.persistState(workingSet.toBytes());
+    return partitionsToConsume;
+  }
```

Comment:
How is the transactional model? If A calls consumePartitions() and B calls it, too, before A commits its transaction, then they will both see the same partitions, right? So we need to document somewhere that the transaction must be committed before you actually start working on the partitions, to find out whether you actually claimed them. Is that documented somewhere?